src/ripple/app/tx/impl/TransactionEngine.cpp

```diff
@@ -133,12 +138,32 @@ TransactionEngine::applyTransaction (
                 }
                 else
                 {
-                    if (fee > balance)
-                        fee = balance;
-                    txnAcct->setFieldAmount (sfBalance, balance - fee);
-                    txnAcct->setFieldU32 (sfSequence, t_seq + 1);
-                    mNodes->entryModify (txnAcct);
-                    didApply = true;
+                    // If the transaction had a Ticket, attempt to consume it.
+                    if ((t_seq == 0) && hasTicket)
+                    {
+                        if (! consumeTicketForTecError (txn,
+                            txn.getFieldAccount160 (sfAccount), this))
+                        {
+                            // Unable to consume Ticket.  Can't charge a 'tec'.
+                            terResult = tefNO_PERMISSION;
```

Comment:
I think you've made a good call here.  I'll try inlining `consumeTicketForTecError()` here.  You can see if you prefer the result.